### PR TITLE
Autocomplete trailing dots

### PR DESCRIPTION
**Pull request by PlasmaIntec**
# Description

Address #6128 by adding the ability to recognize trailing dots in `file_completions.rs`.

Previously, calling `partial_from` on `..` would result in a **base** `.` and a **rest** `..`

This change pushes extraneous dots from **rest** into **base**.

# Tests

Make sure you've done the following:

- [x] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [x] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [x] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [ ] `cargo test --workspace --features=extra` to check that all the tests pass

### CODE REVIEW

**Comment by elferherrera** [comment]
Can you remove the `expect` from your code?

File: crates/nu-cli/src/completions/file_completions.rs [hunk 1]
```diff
@@ -105,6 +105,27 @@ pub fn partial_from(input: &str) -> (String, String) {
...
 
+    // take leftmost dots from rest and push them into base
+    let mut i = 0;
+
+    while i < rest.len() && rest.chars().nth(i).expect("i is within rest length") == '.' {
+        i += 1;
+    }
+
+    if i > 0 {
+        let (leftmost_dots, rest) = rest.split_at(i);
```
Comment (sholderbach) [inline crates/nu-cli/src/completions/file_completions.rs:116]: This code is not Unicode safe this way. `chars` can be more than 1 byte per character.

File: crates/nu-cli/tests/folder_completions.rs [hunk 2]
```diff
@@ -27,3 +29,27 @@ fn folder_completions() {
...
+    let parent_dir = PathBuf::from(dir_str).join("..");
+    let target_dir = format!("cd {}", parent_dir.to_str().unwrap());
+    let suggestions = completer.complete(&target_dir, target_dir.len());
+
+    // Create the expected values
+    let expected_paths: Vec<String> = vec![
+        folder(dir.join("..").join("formats")),
+        folder(dir.join("..").join("playground")),
+        folder(dir.join("..").join("completions")),
+    ];
```
Comment (sholderbach) [inline crates/nu-cli/tests/folder_completions.rs:51]: Can you also add a test case for triggering the completions on a single dot hidden folder or in a case where no initial path is present?

**Review by sholderbach — CHANGES_REQUESTED** [review]
I haven't followed all the code paths, so I can not say definitively that your expects are unreachable. 
Panicking the whole nushell process has to be the last resort. Not doing something in the interactive environment is better than loosing work.
@elferherrera is probably sensitized to `.expect()`s because we had to hotfix panics that resulted from unhandled but possible user inputs in your last PR's.

The logic as it stands has to be changed to account for the fact that there are potentially Unicode characters present. (Maybe you can express that by iterator adapters and use `.char_indices()` or the various string searching/splitting facilities are sufficient to solve that. Also you can [get the unconsumed part of an `Chars` iterator](https://doc.rust-lang.org/std/str/struct.Chars.html#method.as_str))

File: crates/nu-cli/tests/folder_completions.rs (hunk 2 above)
Comment (sholderbach) [reply crates/nu-cli/tests/folder_completions.rs:51]: Our text fixture should probably also include directories with unicode names.

**Comment by PlasmaIntec** [comment]
> I haven't followed all the code paths, so I can not say definitively that your expects are unreachable. Panicking the whole nushell process has to be the last resort. Not doing something in the interactive environment is better than loosing work. @elferherrera is probably sensitized to `.expect()`s because we had to hotfix panics that resulted from unhandled but possible user inputs in your last PR's.
> 
> The logic as it stands has to be changed to account for the fact that there are potentially Unicode characters present. (Maybe you can express that by iterator adapters and use `.char_indices()` or the various string searching/splitting facilities are sufficient to solve that. Also you can [get the unconsumed part of an `Chars` iterator](https://doc.rust-lang.org/std/str/struct.Chars.html#method.as_str))

Do we prefer to simply use `unwrap` over `expect`? This doesn't seem to agree with the code style checker, which gives a:

> error: used `unwrap()` on `an Option` value

**Comment by PlasmaIntec** [comment]
> I haven't followed all the code paths, so I can not say definitively that your expects are unreachable. Panicking the whole nushell process has to be the last resort. Not doing something in the interactive environment is better than loosing work. @elferherrera is probably sensitized to `.expect()`s because we had to hotfix panics that resulted from unhandled but possible user inputs in your last PR's.
> 
> The logic as it stands has to be changed to account for the fact that there are potentially Unicode characters present. (Maybe you can express that by iterator adapters and use `.char_indices()` or the various string searching/splitting facilities are sufficient to solve that. Also you can [get the unconsumed part of an `Chars` iterator](https://doc.rust-lang.org/std/str/struct.Chars.html#method.as_str))

Could you provide a preferable alternative to `.expect()`? I tried both `.unwrap()` and `.unwrap_or()`, but the former goes against the code style and the latter doesn't make much sense with `regex` structs.

**Comment by merelymyself** [comment]
`match` will likely be your friend in this situation.

**Comment by sholderbach** [comment]
Sorry for leaving you hanging for so long. Finally got around to manually testing your PR. 
I like that it implicitly supports the `zsh`-style chaining of more than `..` to get higher up in the hierarchy for each following `.`.

With my initial review I slightly misread what your code is doing as the `while` loop construct with indices was rather verbose and my brain turned unidiomatic into hard to read. Under the given constraints unicode safety in fact shouldn't be a problem, But in general rolling your own indices with UTF-8 encoded strings should only be done with great care, as char index != byte index in general (Thus triggering my mental lint for that.)

For your regex solution the expects should in fact be unreachable, but that is probably out of reach for the compiler to prove so it would generate 3 unnecessary panic handler sections.

I will try to write up a change how to express what you are trying to achieve with a bit more `std` library use that should remove the need to unwrap any values.

There is one rare case to consider: It is legal to have multiple dots at the beginning of a file or name if it contains other valid characters: 
grafik

This would require some changes to emit in this case not only the completions doing the directory traversal but also the regular completion for those rare files. (Bash is able to do that, but doesn't have the leading `./` quirk of nushell to deal with)
I think from a quality of life perspective landing this without getting that solved should still be a massive improvement (but would leave an open issue and a curse for those with `..badfile`)

File: crates/nu-cli/src/completions/file_completions.rs [hunk 3]
```diff
@@ -105,6 +105,19 @@ pub fn partial_from(input: &str) -> (String, String) {
...
+
+    // only push dots from rest into base if we believe input is not a prefix to a single dot hidden folder
+    // TODO: Also deal with legal but rare files or folders that contain multiple leading dots.
+    if num_leading_dots > 1 {
+        base.push_str(leading_dots);
+        base.push(SEP);
+
+        return (base.to_string(), following_leading_dots.to_string());
+    }
+
```
Comment (sholderbach) [inline crates/nu-cli/src/completions/file_completions.rs:120]: Not sure if that logic might have to defer some of that decision to the following file and directoy completers

**Comment by PlasmaIntec** [comment]
Renamed the previous logic worked on into `partial_from_with_parents`, which now returns a `Option<(String, String)>`.

The idea is to let `partial_from` handle the **dots are part of file/directory name** semantics and let `partial_from_with_parents` handle the **dots are parent directory** semantics, then join their results.

**Comment by fdncred** [comment]
moving to draft until we can review this further

**Comment by fdncred** [comment]
Where are we at on this PR?

**Comment by fdncred** [comment]
seems that this isn't moving forward anymore. feel free to submit a new pr with similar functionality based on the conditions stated above. thanks!

**Pull request closed without merging**